src/inmanta/data/__init__.py

```diff
@@ -4815,6 +4815,299 @@ def get_compliance_status(self) -> Optional[state.Compliance]:
+        # Could not express this constraint in the database
+        assert len({rs.name for rs in result}) == len(
+            result
+        ), "Inconsistency in the database, a resource set cannot be present more than once in the same model version"
+        return result
+
+    @classmethod
+    def get_printable_name_for_resource_set(cls, native: str | None) -> str:
+        if native is None:
+            return "<SHARED>"
+        return native
+
+    @classmethod
+    async def validate_resource_sets_in_version(
+        cls,
+        environment: uuid.UUID,
+        version: int,
+        updated_resource_sets: list[str | None],
+        connection: Optional[asyncpg.connection.Connection] = None,
+    ) -> None:
+        """
+        Checks for duplicates in the target version.
+        This should only happen when we try to migrate a resource to another resource set (base -> target)
+        while the base resource set is not present in the partial compile.
+        """
+        query = """
+            WITH all_resources AS (
+                SELECT *
+                FROM resource_set_configuration_model AS rscm
+                INNER JOIN resource_set AS rs
+                    ON rscm.environment=rs.environment
+                    AND rscm.resource_set_id=rs.id
+                INNER JOIN resource AS r
+                    ON rs.environment=r.environment
+                    AND rs.id=r.resource_set_id
+                WHERE rscm.environment=$1
+                    AND rscm.model=$2
+                ),
+                duplicate_resources AS (
+                    SELECT ar.resource_id
+                    FROM all_resources AS ar
+                    GROUP BY ar.resource_id
+                    HAVING COUNT(*) > 1
+                )
+                SELECT ar.resource_id, ar.name, ar.id
+                FROM all_resources AS ar
+                WHERE ar.resource_id=ANY(SELECT * FROM duplicate_resources)
+        """
+        records = await cls._fetch_query(query, environment, version, connection=connection)
+        if records:
+            rid_to_resource_sets: dict[str, dict[str, str]] = {}
+            for record in records:
+                resource_id = str(record["resource_id"])
+                resource_set_name = str(record["name"])
+                if resource_id not in rid_to_resource_sets:
+                    rid_to_resource_sets[resource_id] = {}
+                key = "new" if resource_set_name in updated_resource_sets else "old"
+                if key in rid_to_resource_sets[resource_id]:
+                    raise BadRequest(
+                        f"Resource set with name {resource_set_name} appears more than once on version {version} "
+                        f"of the model with ids {rid_to_resource_sets[resource_id]} and {resource_id}"
+                    )
+                rid_to_resource_sets[resource_id][key] = resource_set_name
```

Comment:
I'm not particularly fond of the old/new keys in a dict. But given that it is contained to this method, I can live with it.